File: backend/src/modules/operatorDefinitions/operatorDefinitions.service.js
const prisma = require('../../lib/prisma');
const { HttpError } = require('../../lib/httpError');
const historian = require('../../lib/historian');

/**
 * Trend definitions/assignments are the sole source of the historian's effective config
 * (coverage + sampleInterval + retentionDays — see historian.js). Every mutation here that could
 * change what a point's enabled/interval/retention/resolvedMappings picture looks like must push
 * a synchronous rebuild before returning, so a caller that immediately re-polls never observes a
 * stale cache — the periodic background refresh in historian.js is a safety net for changes made
 * outside this service, not the primary invalidation path. A refresh failure is logged, never
 * thrown: the Prisma write already succeeded, and the background timer will still catch up.
 * @param {string} kind - 'trend' | 'schedule'; only 'trend' touches the historian cache.
 */
async function refreshHistorianCacheIfTrend(kind) {
  if (kind !== 'trend') return;
  try {
    await historian.refreshHistorianConfig();
  } catch (e) {
    // eslint-disable-next-line no-console
    console.warn('[operatorDefinitions] historian cache refresh failed:', e?.message || e);
  }
}

async function siteOr404(siteId) {
  const site = await prisma.site.findUnique({ where: { id: siteId } });
  if (!site) throw new HttpError(404, 'Site not found');
}

function model(kind) {
  if (kind === 'trend') return { definition: prisma.trendDefinition, assignment: prisma.trendAssignment };
  if (kind === 'schedule') return { definition: prisma.scheduleDefinition, assignment: prisma.scheduleAssignment };
  throw new HttpError(400, 'Unknown definition kind');
}

function validate(kind, body, partial = false) {
  if (!body || typeof body !== 'object' || Array.isArray(body)) throw new HttpError(400, 'Definition must be an object');
  if ((!partial || body.name !== undefined) && (typeof body.name !== 'string' || !body.name.trim())) throw new HttpError(400, 'name is required');
  for (const key of ['enabled', 'isTemplate']) {
    if (body[key] !== undefined && typeof body[key] !== 'boolean') throw new HttpError(400, `${key} must be a boolean`);
  }
  if (kind === 'trend') {
    if (body.sampleInterval != null && (!Number.isInteger(body.sampleInterval) || body.sampleInterval < historian.MIN_SAMPLE_INTERVAL_SECONDS)) throw new HttpError(400, `sampleInterval must be an integer of at least ${historian.MIN_SAMPLE_INTERVAL_SECONDS} seconds`);
    if (body.pointRequirements !== undefined && (!Array.isArray(body.pointRequirements) || body.pointRequirements.some((point) => !point || typeof point.pointKey !== 'string' || !point.pointKey.trim()))) throw new HttpError(400, 'pointRequirements must contain logical point keys');
  } else if (body.weeklyWindows !== undefined) {
    const days = ['Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat', 'Sun'];
    const time = /^([01]\d|2[0-3]):[0-5]\d$/;
    if (!Array.isArray(body.weeklyWindows) || body.weeklyWindows.some((window) => !window || !Array.isArray(window.days) || !window.days.length || window.days.some((day) => !days.includes(day)) || !time.test(window.startTime) || !time.test(window.endTime) || !['Occupied', 'Unoccupied'].includes(window.action))) throw new HttpError(400, 'weeklyWindows must contain valid days, times and occupancy actions');
  }
}

async function list(siteId, kind, query = {}) {
  await siteOr404(siteId); const m = model(kind);
  return m.definition.findMany({ where: { siteId, ...(query.templates === 'true' ? { isTemplate: true } : {}), ...(query.templates === 'false' ? { isTemplate: false } : {}) }, include: { assignments: true }, orderBy: { updatedAt: 'desc' } });
}

async function create(siteId, kind, body) {
  validate(kind, body);
  await siteOr404(siteId); const m = model(kind);
  const data = kind === 'trend'
    ? { siteId, name: String(body.name || '').trim(), enabled: body.enabled !== false, isTemplate: Boolean(body.isTemplate), equipmentType: body.equipmentType || null, sampleInterval: body.sampleInterval == null ? null : Number(body.sampleInterval), pointRequirements: body.pointRequirements || [], ...(body.retentionDays != null ? { retentionDays: Number(body.retentionDays) } : {}) }
    : { siteId, name: String(body.name || '').trim(), enabled: body.enabled !== false, isTemplate: Boolean(body.isTemplate), weeklyWindows: body.weeklyWindows || [] };
  if (!data.name) throw new HttpError(400, 'name is required');
  // Templates are global, reusable definitions (Save as Template is a distinct, explicit action
  // from a per-equipment Save). A same-named template is never silently overwritten or duplicated
  // here — the caller must pick a different name, or update/version the existing template instead.
  if (data.isTemplate) {
    const dup = await m.definition.findFirst({
      where: { siteId, isTemplate: true, name: { equals: data.name, mode: 'insensitive' } },
      select: { id: true },
    });
    if (dup) throw new HttpError(409, `A template named "${data.name}" already exists`);
  }
  if (body.equipmentIds !== undefined) {
    if (!Array.isArray(body.equipmentIds) || !body.equipmentIds.length) throw new HttpError(400, 'equipmentIds must be a nonempty array');
    const ids = [...new Set(body.equipmentIds.map(String))];
    const equipment = await prisma.equipment.findMany({ where: { siteId, id: { in: ids } }, select: { id: true } });
    if (equipment.length !== ids.length) throw new HttpError(400, 'Every assignment must reference equipment in this site');
    data.assignments = { create: ids.map((equipmentId) => ({ siteId, equipmentId, ...(kind === 'trend' ? { resolvedMappings: body.resolvedMappings?.[equipmentId] || {} } : {}) })) };
  }
  const created = await m.definition.create({ data, include: { assignments: true } });
  await refreshHistorianCacheIfTrend(kind);
  return created;
}

async function update(siteId, kind, id, body) {
  validate(kind, body, true);
  const m = model(kind); const existing = await m.definition.findFirst({ where: { id, siteId } });
  if (!existing) throw new HttpError(404, 'Definition not found');
  const data = { ...(body.name !== undefined ? { name: String(body.name).trim() } : {}), ...(body.enabled !== undefined ? { enabled: Boolean(body.enabled) } : {}), ...(kind === 'trend' ? { ...(body.pointRequirements !== undefined ? { pointRequirements: body.pointRequirements } : {}), ...(body.sampleInterval !== undefined ? { sampleInterval: body.sampleInterval == null ? null : Number(body.sampleInterval) } : {}) } : { ...(body.weeklyWindows !== undefined ? { weeklyWindows: body.weeklyWindows } : {}) }), version: { increment: 1 } };
  if (body.isTemplate !== undefined) data.isTemplate = body.isTemplate;
  if (kind === 'trend' && body.equipmentType !== undefined) data.equipmentType = body.equipmentType || null;
  if (kind === 'trend' && body.retentionDays !== undefined) data.retentionDays = body.retentionDays == null ? null : Number(body.retentionDays);
  // Same duplicate-name guard as create(), applied whenever the row will be (or already is) a
  // template and its name is changing: renaming a template, or promoting a per-equipment
  // definition into a template, must not collide with an existing template name.
  const willBeTemplate = data.isTemplate !== undefined ? data.isTemplate : existing.isTemplate;
  if (willBeTemplate && data.name !== undefined && data.name.toLowerCase() !== String(existing.name || '').toLowerCase()) {
    const dup = await m.definition.findFirst({
      where: { siteId, isTemplate: true, name: { equals: data.name, mode: 'insensitive' }, id: { not: id } },
      select: { id: true },
    });
    if (dup) throw new HttpError(409, `A template named "${data.name}" already exists`);
  }
  const updated = await m.definition.update({ where: { id }, data, include: { assignments: true } });
  await refreshHistorianCacheIfTrend(kind);
  return updated;
}

async function remove(siteId, kind, id) {
  const m = model(kind);
  const existing = await m.definition.findFirst({ where: { id, siteId } });
  if (!existing) throw new HttpError(404, 'Definition not found');
  await m.definition.delete({ where: { id } });
  await refreshHistorianCacheIfTrend(kind);
  return { ok: true };
}

async function assign(siteId, kind, id, body) {
  if (!body || (body.equipmentIds !== undefined && !Array.isArray(body.equipmentIds))) throw new HttpError(400, 'equipmentIds must be an array');
  const m = model(kind); const definition = await m.definition.findFirst({ where: { id, siteId } }); if (!definition) throw new HttpError(404, 'Definition not found');
  const equipmentIds = [...new Set((body.equipmentIds || (body.equipmentId ? [body.equipmentId] : [])).map(String))];
  if (!equipmentIds.length) throw new HttpError(400, 'equipmentIds are required');
  const equipment = await prisma.equipment.findMany({ where: { id: { in: equipmentIds }, siteId }, select: { id: true } });
  if (equipment.length !== equipmentIds.length) throw new HttpError(400, 'Every assignment must reference equipment in this site');
  const operations = [];
  for (const equipmentId of equipmentIds) {
    const data = { enabled: body.enabled !== false, ...(kind === 'trend' ? { resolvedMappings: body.resolvedMappings?.[equipmentId] || {} } : {}) };
    operations.push(m.assignment.upsert({ where: { definitionId_equipmentId: { definitionId: id, equipmentId } }, create: { definitionId: id, siteId, equipmentId, ...data }, update: data }));
  }
  const result = await prisma.$transaction(operations);
  await refreshHistorianCacheIfTrend(kind);
  return result;
}

async function unassign(siteId, kind, id, equipmentId) {
  const m = model(kind);
  await m.assignment.deleteMany({ where: { definitionId: id, siteId, equipmentId } });
  await refreshHistorianCacheIfTrend(kind);
  return { ok: true };
}

module.exports = { list, create, update, remove, assign, unassign };

File: server/src/main/java/com/legioncontrols/server/operator/OperatorDefinitionsService.java
package com.legioncontrols.server.operator;

import com.legioncontrols.server.common.ApiException;
import com.legioncontrols.server.common.JsonUtil;
import com.legioncontrols.server.historian.HistorianService;
import com.legioncontrols.server.jooq.generated.tables.records.EquipmentRecord;
import com.legioncontrols.server.jooq.generated.tables.records.ScheduleassignmentRecord;
import com.legioncontrols.server.jooq.generated.tables.records.ScheduledefinitionRecord;
import com.legioncontrols.server.jooq.generated.tables.records.TrendassignmentRecord;
import com.legioncontrols.server.jooq.generated.tables.records.TrenddefinitionRecord;
import java.time.LocalDateTime;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.regex.Pattern;
import org.jooq.DSLContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;
import tools.jackson.databind.JsonNode;
import tools.jackson.databind.node.ArrayNode;
import tools.jackson.databind.node.ObjectNode;

import static com.legioncontrols.server.jooq.generated.tables.Equipment.EQUIPMENT;
import static com.legioncontrols.server.jooq.generated.tables.Scheduleassignment.SCHEDULEASSIGNMENT;
import static com.legioncontrols.server.jooq.generated.tables.Scheduledefinition.SCHEDULEDEFINITION;
import static com.legioncontrols.server.jooq.generated.tables.Site.SITE;
import static com.legioncontrols.server.jooq.generated.tables.Trendassignment.TRENDASSIGNMENT;
import static com.legioncontrols.server.jooq.generated.tables.Trenddefinition.TRENDDEFINITION;

/**
 * Java equivalent of backend/src/modules/operatorDefinitions/operatorDefinitions.service.js — a
 * generic definition/assignment CRUD shared by trend and schedule "kinds". Trend definitions are
 * the sole source of HistorianService's effective config (coverage + sampleInterval +
 * retentionDays), so every trend mutation here synchronously refreshes that cache before
 * returning — the same invariant as the Node version.
 */
@Service
public class OperatorDefinitionsService {

    private static final Logger log = LoggerFactory.getLogger(OperatorDefinitionsService.class);
    private static final int MIN_SAMPLE_INTERVAL_SECONDS = 10;
    private static final Set<String> WEEKDAYS = Set.of("Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun");
    private static final Pattern TIME_PATTERN = Pattern.compile("^([01]\\d|2[0-3]):[0-5]\\d$");

    public enum Kind { TREND, SCHEDULE }

    public static Kind parseKind(String kind) {
        if ("trend".equals(kind)) return Kind.TREND;
        if ("schedule".equals(kind)) return Kind.SCHEDULE;
        throw ApiException.badRequest("Unknown definition kind");
    }

    private final DSLContext dsl;
    private final JsonUtil json;
    private final HistorianService historianService;

    public OperatorDefinitionsService(DSLContext dsl, JsonUtil json, HistorianService historianService) {
        this.dsl = dsl;
        this.json = json;
        this.historianService = historianService;
    }

    private void assertSite(String siteId) {
        boolean exists = dsl.fetchExists(dsl.selectFrom(SITE).where(SITE.ID.eq(siteId)));
        if (!exists) throw ApiException.notFound("Site not found");
    }

    private void refreshHistorianCacheIfTrend(Kind kind) {
        if (kind != Kind.TREND) return;
        try {
            historianService.refreshHistorianConfig();
        } catch (Exception e) {
            log.warn("[operatorDefinitions] historian cache refresh failed: {}", e.getMessage());
        }
    }

    // ---- validation --------------------------------------------------------------------------

    private void validate(Kind kind, JsonNode body, boolean partial) {
        if (body == null || !body.isObject()) throw ApiException.badRequest("Definition must be an object");
        boolean nameRequired = !partial || body.has("name");
        if (nameRequired && (!body.has("name") || !body.get("name").isTextual() || body.get("name").asText().isBlank())) {
            throw ApiException.badRequest("name is required");
        }
        for (String key : new String[]{"enabled", "isTemplate"}) {
            if (body.has(key) && !body.get(key).isBoolean()) {
                throw ApiException.badRequest(key + " must be a boolean");
            }
        }
        if (kind == Kind.TREND) {
            if (body.has("sampleInterval") && !body.get("sampleInterval").isNull()) {
                JsonNode si = body.get("sampleInterval");
                if (!si.isIntegralNumber() || si.asInt() < MIN_SAMPLE_INTERVAL_SECONDS) {
                    throw ApiException.badRequest("sampleInterval must be an integer of at least " + MIN_SAMPLE_INTERVAL_SECONDS + " seconds");
                }
            }
            if (body.has("pointRequirements")) {
                JsonNode pr = body.get("pointRequirements");
                boolean valid = pr.isArray();
                if (valid) {
                    for (JsonNode point : pr) {
                        if (point == null || !point.isObject() || !point.hasNonNull("pointKey") || point.get("pointKey").asText().isBlank()) {
                            valid = false;
                            break;
                        }
                    }
                }
                if (!valid) throw ApiException.badRequest("pointRequirements must contain logical point keys");
            }
        } else if (body.has("weeklyWindows")) {
            JsonNode windows = body.get("weeklyWindows");
            boolean valid = windows.isArray();
            if (valid) {
                for (JsonNode window : windows) {
                    valid = window != null && window.isObject()
                        && window.get("days") != null && window.get("days").isArray() && !window.get("days").isEmpty()
                        && allDaysValid(window.get("days"))
                        && window.hasNonNull("startTime") && TIME_PATTERN.matcher(window.get("startTime").asText()).matches()
                        && window.hasNonNull("endTime") && TIME_PATTERN.matcher(window.get("endTime").asText()).matches()
                        && window.hasNonNull("action") && ("Occupied".equals(window.get("action").asText()) || "Unoccupied".equals(window.get("action").asText()));
                    if (!valid) break;
                }
            }
            if (!valid) throw ApiException.badRequest("weeklyWindows must contain valid days, times and occupancy actions");
        }
    }

    private static boolean allDaysValid(JsonNode days) {
        for (JsonNode day : days) {
            if (!WEEKDAYS.contains(day.asText())) return false;
        }
        return true;
    }

    // ---- list ----------------------------------------------------------------------------------

    public Object list(String siteId, Kind kind, String templatesFilter) {
        assertSite(siteId);
        if (kind == Kind.TREND) {
            var condition = TRENDDEFINITION.SITEID.eq(siteId);
            var step = dsl.selectFrom(TRENDDEFINITION).where(condition);
            List<TrenddefinitionRecord> defs = "true".equals(templatesFilter)
                ? dsl.selectFrom(TRENDDEFINITION).where(condition, TRENDDEFINITION.ISTEMPLATE.isTrue()).orderBy(TRENDDEFINITION.UPDATEDAT.desc()).fetch()
                : "false".equals(templatesFilter)
                    ? dsl.selectFrom(TRENDDEFINITION).where(condition, TRENDDEFINITION.ISTEMPLATE.isFalse()).orderBy(TRENDDEFINITION.UPDATEDAT.desc()).fetch()
                    : step.orderBy(TRENDDEFINITION.UPDATEDAT.desc()).fetch();
            return defs.stream().map(d -> TrendDefinitionDto.from(d, fetchTrendAssignments(d.getId()), json)).toList();
        } else {
            var condition = SCHEDULEDEFINITION.SITEID.eq(siteId);
            List<ScheduledefinitionRecord> defs = "true".equals(templatesFilter)
                ? dsl.selectFrom(SCHEDULEDEFINITION).where(condition, SCHEDULEDEFINITION.ISTEMPLATE.isTrue()).orderBy(SCHEDULEDEFINITION.UPDATEDAT.desc()).fetch()
                : "false".equals(templatesFilter)
                    ? dsl.selectFrom(SCHEDULEDEFINITION).where(condition, SCHEDULEDEFINITION.ISTEMPLATE.isFalse()).orderBy(SCHEDULEDEFINITION.UPDATEDAT.desc()).fetch()
                    : dsl.selectFrom(SCHEDULEDEFINITION).where(condition).orderBy(SCHEDULEDEFINITION.UPDATEDAT.desc()).fetch();
            return defs.stream().map(d -> ScheduleDefinitionDto.from(d, fetchScheduleAssignments(d.getId()), json)).toList();
        }
    }

    private List<TrendassignmentRecord> fetchTrendAssignments(String definitionId) {
        return dsl.selectFrom(TRENDASSIGNMENT).where(TRENDASSIGNMENT.DEFINITIONID.eq(definitionId)).fetch();
    }

    private List<ScheduleassignmentRecord> fetchScheduleAssignments(String definitionId) {
        return dsl.selectFrom(SCHEDULEASSIGNMENT).where(SCHEDULEASSIGNMENT.DEFINITIONID.eq(definitionId)).fetch();
    }

    // ---- create ----------------------------------------------------------------------------------

    public Object create(String siteId, Kind kind, JsonNode body) {
        validate(kind, body, false);
        assertSite(siteId);
        String name = body.get("name").asText().trim();
        boolean enabled = !body.has("enabled") || body.get("enabled").asBoolean();
        boolean isTemplate = body.has("isTemplate") && body.get("isTemplate").asBoolean();

        if (isTemplate) {
            boolean dup = kind == Kind.TREND
                ? dsl.fetchExists(dsl.selectFrom(TRENDDEFINITION).where(TRENDDEFINITION.SITEID.eq(siteId), TRENDDEFINITION.ISTEMPLATE.isTrue(), TRENDDEFINITION.NAME.equalIgnoreCase(name)))
                : dsl.fetchExists(dsl.selectFrom(SCHEDULEDEFINITION).where(SCHEDULEDEFINITION.SITEID.eq(siteId), SCHEDULEDEFINITION.ISTEMPLATE.isTrue(), SCHEDULEDEFINITION.NAME.equalIgnoreCase(name)));
            if (dup) throw ApiException.conflict("A template named \"" + name + "\" already exists");
        }

        List<String> equipmentIds = null;
        if (body.has("equipmentIds")) {
            JsonNode idsNode = body.get("equipmentIds");
            if (!idsNode.isArray() || idsNode.isEmpty()) throw ApiException.badRequest("equipmentIds must be a nonempty array");
            Set<String> ids = new LinkedHashSet<>();
            for (JsonNode id : idsNode) ids.add(id.asText());
            equipmentIds = List.copyOf(ids);
            long matchCount = dsl.selectCount().from(EQUIPMENT).where(EQUIPMENT.SITEID.eq(siteId), EQUIPMENT.ID.in(equipmentIds)).fetchOne(0, Long.class);
            if (matchCount != equipmentIds.size()) throw ApiException.badRequest("Every assignment must reference equipment in this site");
        }

        LocalDateTime now = LocalDateTime.now();
        Object result;
        if (kind == Kind.TREND) {
            TrenddefinitionRecord record = dsl.newRecord(TRENDDEFINITION);
            record.setId(UUID.randomUUID().toString());
            record.setSiteid(siteId);
            record.setName(name);
            record.setEnabled(enabled);
            record.setIstemplate(isTemplate);
            record.setEquipmenttype(body.has("equipmentType") && !body.get("equipmentType").isNull() ? body.get("equipmentType").asText() : null);
            record.setSampleinterval(body.has("sampleInterval") && !body.get("sampleInterval").isNull() ? body.get("sampleInterval").asInt() : null);
            record.setPointrequirements(json.toJsonb(body.has("pointRequirements") ? body.get("pointRequirements") : json.newArray()));
            if (body.has("retentionDays") && !body.get("retentionDays").isNull()) record.setRetentiondays(body.get("retentionDays").asInt());
            record.setCreatedat(now);
            record.setUpdatedat(now);
            record.insert();

            if (equipmentIds != null) {
                JsonNode resolvedMappings = body.get("resolvedMappings");
                for (String equipmentId : equipmentIds) {
                    TrendassignmentRecord assignment = dsl.newRecord(TRENDASSIGNMENT);
                    assignment.setId(UUID.randomUUID().toString());
                    assignment.setDefinitionid(record.getId());
                    assignment.setSiteid(siteId);
                    assignment.setEquipmentid(equipmentId);
                    assignment.setEnabled(true);
                    JsonNode mapping = resolvedMappings != null ? resolvedMappings.get(equipmentId) : null;
                    assignment.setResolvedmappings(json.toJsonb(mapping != null ? mapping : json.newObject()));
                    assignment.setCreatedat(now);
                    assignment.setUpdatedat(now);
                    assignment.insert();
                }
            }
            record.refresh();
            result = TrendDefinitionDto.from(record, fetchTrendAssignments(record.getId()), json);
        } else {
            ScheduledefinitionRecord record = dsl.newRecord(SCHEDULEDEFINITION);
            record.setId(UUID.randomUUID().toString());
            record.setSiteid(siteId);
            record.setName(name);
            record.setEnabled(enabled);
            record.setIstemplate(isTemplate);
            record.setWeeklywindows(json.toJsonb(body.has("weeklyWindows") ? body.get("weeklyWindows") : json.newArray()));
            record.setCreatedat(now);
            record.setUpdatedat(now);
            record.insert();

            if (equipmentIds != null) {
                for (String equipmentId : equipmentIds) {
                    ScheduleassignmentRecord assignment = dsl.newRecord(SCHEDULEASSIGNMENT);
                    assignment.setId(UUID.randomUUID().toString());
                    assignment.setDefinitionid(record.getId());
                    assignment.setSiteid(siteId);
                    assignment.setEquipmentid(equipmentId);
                    assignment.setEnabled(true);
                    assignment.setCreatedat(now);
                    assignment.setUpdatedat(now);
                    assignment.insert();
                }
            }
            record.refresh();
            result = ScheduleDefinitionDto.from(record, fetchScheduleAssignments(record.getId()), json);
        }

        refreshHistorianCacheIfTrend(kind);
        return result;
    }

    // ---- update ----------------------------------------------------------------------------------

    public Object update(String siteId, Kind kind, String id, JsonNode body) {
        validate(kind, body, true);
        Object result;
        if (kind == Kind.TREND) {
            TrenddefinitionRecord existing = dsl.selectFrom(TRENDDEFINITION).where(TRENDDEFINITION.ID.eq(id), TRENDDEFINITION.SITEID.eq(siteId)).fetchOne();
            if (existing == null) throw ApiException.notFound("Definition not found");
            String newName = body.has("name") ? body.get("name").asText().trim() : existing.getName();
            boolean willBeTemplate = body.has("isTemplate") ? body.get("isTemplate").asBoolean() : Boolean.TRUE.equals(existing.getIstemplate());
            if (willBeTemplate && body.has("name") && !newName.equalsIgnoreCase(existing.getName())) {
                boolean dup = dsl.fetchExists(dsl.selectFrom(TRENDDEFINITION)
                    .where(TRENDDEFINITION.SITEID.eq(siteId), TRENDDEFINITION.ISTEMPLATE.isTrue(), TRENDDEFINITION.NAME.equalIgnoreCase(newName), TRENDDEFINITION.ID.ne(id)));
                if (dup) throw ApiException.conflict("A template named \"" + newName + "\" already exists");
            }
            if (body.has("name")) existing.setName(newName);
            if (body.has("enabled")) existing.setEnabled(body.get("enabled").asBoolean());
            if (body.has("isTemplate")) existing.setIstemplate(body.get("isTemplate").asBoolean());
            if (body.has("pointRequirements")) existing.setPointrequirements(json.toJsonb(body.get("pointRequirements")));
            if (body.has("sampleInterval")) existing.setSampleinterval(body.get("sampleInterval").isNull() ? null : body.get("sampleInterval").asInt());
            if (body.has("equipmentType")) existing.setEquipmenttype(body.get("equipmentType").isNull() || body.get("equipmentType").asText().isBlank() ? null : body.get("equipmentType").asText());
            if (body.has("retentionDays")) existing.setRetentiondays(body.get("retentionDays").isNull() ? null : body.get("retentionDays").asInt());
            existing.setVersion(existing.getVersion() + 1);
            existing.setUpdatedat(LocalDateTime.now());
            existing.update();
            result = TrendDefinitionDto.from(existing, fetchTrendAssignments(id), json);
        } else {
            ScheduledefinitionRecord existing = dsl.selectFrom(SCHEDULEDEFINITION).where(SCHEDULEDEFINITION.ID.eq(id), SCHEDULEDEFINITION.SITEID.eq(siteId)).fetchOne();
            if (existing == null) throw ApiException.notFound("Definition not found");
            String newName = body.has("name") ? body.get("name").asText().trim() : existing.getName();
            boolean willBeTemplate = body.has("isTemplate") ? body.get("isTemplate").asBoolean() : Boolean.TRUE.equals(existing.getIstemplate());
            if (willBeTemplate && body.has("name") && !newName.equalsIgnoreCase(existing.getName())) {
                boolean dup = dsl.fetchExists(dsl.selectFrom(SCHEDULEDEFINITION)
                    .where(SCHEDULEDEFINITION.SITEID.eq(siteId), SCHEDULEDEFINITION.ISTEMPLATE.isTrue(), SCHEDULEDEFINITION.NAME.equalIgnoreCase(newName), SCHEDULEDEFINITION.ID.ne(id)));
                if (dup) throw ApiException.conflict("A template named \"" + newName + "\" already exists");
            }
            if (body.has("name")) existing.setName(newName);
            if (body.has("enabled")) existing.setEnabled(body.get("enabled").asBoolean());
            if (body.has("isTemplate")) existing.setIstemplate(body.get("isTemplate").asBoolean());
            if (body.has("weeklyWindows")) existing.setWeeklywindows(json.toJsonb(body.get("weeklyWindows")));
            existing.setVersion(existing.getVersion() + 1);
            existing.setUpdatedat(LocalDateTime.now());
            existing.update();
            result = ScheduleDefinitionDto.from(existing, fetchScheduleAssignments(id), json);
        }
        refreshHistorianCacheIfTrend(kind);
        return result;
    }

    // ---- remove ----------------------------------------------------------------------------------

    public void remove(String siteId, Kind kind, String id) {
        if (kind == Kind.TREND) {
            boolean exists = dsl.fetchExists(dsl.selectFrom(TRENDDEFINITION).where(TRENDDEFINITION.ID.eq(id), TRENDDEFINITION.SITEID.eq(siteId)));
            if (!exists) throw ApiException.notFound("Definition not found");
            dsl.deleteFrom(TRENDDEFINITION).where(TRENDDEFINITION.ID.eq(id)).execute();
        } else {
            boolean exists = dsl.fetchExists(dsl.selectFrom(SCHEDULEDEFINITION).where(SCHEDULEDEFINITION.ID.eq(id), SCHEDULEDEFINITION.SITEID.eq(siteId)));
            if (!exists) throw ApiException.notFound("Definition not found");
            dsl.deleteFrom(SCHEDULEDEFINITION).where(SCHEDULEDEFINITION.ID.eq(id)).execute();
        }
        refreshHistorianCacheIfTrend(kind);
    }

    // ---- assign / unassign -----------------------------------------------------------------------

    public Object assign(String siteId, Kind kind, String id, JsonNode body) {
        if (body != null && body.has("equipmentIds") && !body.get("equipmentIds").isArray()) {
            throw ApiException.badRequest("equipmentIds must be an array");
        }
        Set<String> equipmentIds = new LinkedHashSet<>();
        if (body != null && body.has("equipmentIds")) {
            for (JsonNode id2 : body.get("equipmentIds")) equipmentIds.add(id2.asText());
        } else if (body != null && body.has("equipmentId")) {
            equipmentIds.add(body.get("equipmentId").asText());
        }
        if (equipmentIds.isEmpty()) throw ApiException.badRequest("equipmentIds are required");

        long matchCount = dsl.selectCount().from(EQUIPMENT).where(EQUIPMENT.ID.in(equipmentIds), EQUIPMENT.SITEID.eq(siteId)).fetchOne(0, Long.class);
        if (matchCount != equipmentIds.size()) throw ApiException.badRequest("Every assignment must reference equipment in this site");

        boolean enabled = body == null || !body.has("enabled") || body.get("enabled").asBoolean();
        LocalDateTime now = LocalDateTime.now();
        Object result;

        if (kind == Kind.TREND) {
            boolean defExists = dsl.fetchExists(dsl.selectFrom(TRENDDEFINITION).where(TRENDDEFINITION.ID.eq(id), TRENDDEFINITION.SITEID.eq(siteId)));
            if (!defExists) throw ApiException.notFound("Definition not found");
            JsonNode resolvedMappings = body != null ? body.get("resolvedMappings") : null;
            List<TrendassignmentRecord> saved = dsl.transactionResult(cfg -> {
                var tx = org.jooq.impl.DSL.using(cfg);
                List<TrendassignmentRecord> out = new java.util.ArrayList<>();
                for (String equipmentId : equipmentIds) {
                    JsonNode mapping = resolvedMappings != null ? resolvedMappings.get(equipmentId) : null;
                    TrendassignmentRecord existing = tx.selectFrom(TRENDASSIGNMENT)
                        .where(TRENDASSIGNMENT.DEFINITIONID.eq(id), TRENDASSIGNMENT.EQUIPMENTID.eq(equipmentId)).fetchOne();
                    if (existing != null) {
                        existing.setEnabled(enabled);
                        existing.setResolvedmappings(json.toJsonb(mapping != null ? mapping : json.newObject()));
                        existing.setUpdatedat(now);
                        existing.update();
                        out.add(existing);
                    } else {
                        TrendassignmentRecord created = tx.newRecord(TRENDASSIGNMENT);
                        created.setId(UUID.randomUUID().toString());
                        created.setDefinitionid(id);
                        created.setSiteid(siteId);
                        created.setEquipmentid(equipmentId);
                        created.setEnabled(enabled);
                        created.setResolvedmappings(json.toJsonb(mapping != null ? mapping : json.newObject()));
                        created.setCreatedat(now);
                        created.setUpdatedat(now);
                        created.insert();
                        out.add(created);
                    }
                }
                return out;
            });
            result = saved.stream().map(a -> TrendDefinitionDto.TrendAssignmentDto.from(a, json)).toList();
        } else {
            boolean defExists = dsl.fetchExists(dsl.selectFrom(SCHEDULEDEFINITION).where(SCHEDULEDEFINITION.ID.eq(id), SCHEDULEDEFINITION.SITEID.eq(siteId)));
            if (!defExists) throw ApiException.notFound("Definition not found");
            List<ScheduleassignmentRecord> saved = dsl.transactionResult(cfg -> {
                var tx = org.jooq.impl.DSL.using(cfg);
                List<ScheduleassignmentRecord> out = new java.util.ArrayList<>();
                for (String equipmentId : equipmentIds) {
                    ScheduleassignmentRecord existing = tx.selectFrom(SCHEDULEASSIGNMENT)
                        .where(SCHEDULEASSIGNMENT.DEFINITIONID.eq(id), SCHEDULEASSIGNMENT.EQUIPMENTID.eq(equipmentId)).fetchOne();
                    if (existing != null) {
                        existing.setEnabled(enabled);
                        existing.setUpdatedat(now);
                        existing.update();
                        out.add(existing);
                    } else {
                        ScheduleassignmentRecord created = tx.newRecord(SCHEDULEASSIGNMENT);
                        created.setId(UUID.randomUUID().toString());
                        created.setDefinitionid(id);
                        created.setSiteid(siteId);
                        created.setEquipmentid(equipmentId);
                        created.setEnabled(enabled);
                        created.setCreatedat(now);
                        created.setUpdatedat(now);
                        created.insert();
                        out.add(created);
                    }
                }
                return out;
            });
            result = saved.stream().map(ScheduleDefinitionDto.ScheduleAssignmentDto::from).toList();
        }

        refreshHistorianCacheIfTrend(kind);
        return result;
    }

    public void unassign(String siteId, Kind kind, String id, String equipmentId) {
        if (kind == Kind.TREND) {
            dsl.deleteFrom(TRENDASSIGNMENT)
                .where(TRENDASSIGNMENT.DEFINITIONID.eq(id), TRENDASSIGNMENT.SITEID.eq(siteId), TRENDASSIGNMENT.EQUIPMENTID.eq(equipmentId))
                .execute();
        } else {
            dsl.deleteFrom(SCHEDULEASSIGNMENT)
                .where(SCHEDULEASSIGNMENT.DEFINITIONID.eq(id), SCHEDULEASSIGNMENT.SITEID.eq(siteId), SCHEDULEASSIGNMENT.EQUIPMENTID.eq(equipmentId))
                .execute();
        }
        refreshHistorianCacheIfTrend(kind);
    }
}
